### DISCUSSION TITLE
Data-only message not received on foreground on iOS

### DISCUSSION THREAD

**Discussion by lucasfronza** [invertase/react-native-firebase#4614] [Q&A]
Hey! 👋 
First of all, thanks very much for taking the time to read and reply!

I've seen plenty of people having trouble receiving data-only messages while the app is on background or quit states. My case is the very opposite. We don't care for background messages, just for foreground ones. So everything from now on is considering that the app is on foreground.

For Android, both "notification + data" and "data-only" messages are picked up by the `.messaging().onMessage(...)` listener. For iOS, only "notification + data" is.

Some questions that I have:
1- How to debug if the message is received by the iOS device? Like using `adb` for Android? Or since it is a data-only message, it is handled directly by the JS code and does not go through the native side? (we are not using `content-available: true`)

2- I added many breakpoints on all the native `didReceiveRemoteNotification` and `willPresentNotification` methods that I could find, but they are only "called" when using "notification + data" messages. Which methods should the data-only messages go through? Where to look for?

Some extra info:
- It worked when we were using v5, and stopped working right after we upgraded it to the versions below. I double-checked all the configs, I might have missed something, but want to find out where to look for. 
```sh
├─ @react-native-firebase/analytics@7.6.10
├─ @react-native-firebase/app@8.4.3
├─ @react-native-firebase/auth@9.2.3
├─ @react-native-firebase/dynamic-links@7.5.4
├─ @react-native-firebase/iid@7.4.4
├─ @react-native-firebase/messaging@7.8.6
├─ @react-native-firebase/remote-config@9.0.5
```
- Tried upgrading everything to v10, it didn't fix the issue.
- Swizzling is not disabled
- Other installed packages: `@react-native-community/push-notification-ios`, `react-native-push-notification`

Thanks!

**Comment by andersonaddo** [comment]
Yeah, there is something like adb for ios. 
If you're testing notifications on a real device (which, of course, you should be), then you can plug it into a computer and see logs on Xcode. I would suggest looking into that first.

Also, what does the message you're sending look like? Please show us.
Also, this may be a apns configuration issue; maybe you used the wrong headers or something. Have a look at that too :)

**Reply by lucasfronza** [reply]
I don't think it is an apns config issue because if I build our app version with firebase v5, it works for both "notification + data" and "data-only" messages. 🤔 

We use `POST https://fcm.googleapis.com/fcm/send`,
```python
# headers
{
    "Content-Type": "Application/json",
    "Authorization": "key={}".format(api_key),
    "project_id": sender_id,
}

# data
{
    "to": ...,
    "data": {
        "some": "value",
    },
    # notification + data
    "notification": { 
        "title": "Some title",
        "body": "Some body",
    },
}
```

**Reply by mikehardy** [reply]
missing content-available I think maybe still necessary for foreground, also you can try adding `-FIRMessagesDebugEnabled` argument to XCode schema, it's a recent addition that helps figure these things out.

Make absolutely sure you are on v10 of everything. With underlying firebase-ios-sdk of v7.1.0. Make sure you are using a real device.

**Reply by lucasfronza** [reply]
Using a real device, yeah. Tried multiple variants adding and changing `content-available`, `priority` and `ttl` settings. It didn't help 😢 . I'll definitely try `-FIRMessagesDebugEnabled`. And will bump v10 as soon as possible 👍 . Upgrade on iOS was smooth, but not on Android (Issue with other packages using different google play versions I believe. Didn't spend much time on it yet since it didn't fix the iOS issue).

**Reply by mikehardy** [reply]
for iOS there is a specific problem where notifications were not received until app restart because of a token registration issue in firebase-ios-sdk 6.34.0 at least (maybe 6.33.0 as well) and was fixed in v7.0.0. v7.1.0 is current. So you definitely want fresh versions

**Reply by lucasfronza** [reply]
Ok. Good to know!

**Reply by lucasfronza** [reply]
Will follow the recommendations and hopefully it will fix it. Thanks both! 🙏
